<a href="https://colab.research.google.com/github/mariea-aashif/Statistical-Learning-e22374/blob/main/Assignment_3_Conditional_Expectation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## ASSIGNMENT 3 : CONDITIONAL EXPECTATION
### E/22/374 - M.A.M. SHIFAA
### QUESTION 1
### 1)
The probability space is

$$
\Omega=\{\omega_1,\omega_2,\omega_3,\omega_4\},
$$

$$
\mathfrak F=\mathfrak P(\Omega),
$$

and the probability measure $P$ is defined by

$$
P(\{\omega_1\})=0.50,\qquad
P(\{\omega_2\})=0.10,\qquad
P(\{\omega_3\})=0.20,\qquad
P(\{\omega_4\})=0.20.
$$

Since

$$
0.50+0.10+0.20+0.20=1,
$$

this is a valid probability measure.

---

### 2)
The random variable $Y$ is defined by

$$
Y(\omega_1)=0,\qquad
Y(\omega_2)=0,\qquad
Y(\omega_3)=1,\qquad
Y(\omega_4)=1.
$$

Thus the inverse images of the values of $Y$ are

$$
Y^{-1}(\{0\})=\{\omega_1,\omega_2\},
$$

and

$$
Y^{-1}(\{1\})=\{\omega_3,\omega_4\}.
$$

Therefore the $\sigma$-algebra generated by $Y$ is

$$
\sigma(Y)=\left\{\varnothing,\Omega,\{\omega_1,\omega_2\},\{\omega_3,\omega_4\}\right\}.
$$

---

### 3)
By definition, $\mathbb{E}[X\mid Y]$ must be $\sigma(Y)$-measurable. This structural constraint means that the estimator can only depend on the macro-information revealed by $Y$, rather than the fine-grained individual outcomes of $\Omega$.

Since $\sigma(Y)$ cannot distinguish between outcomes within the partition cells

$$
\{\omega_1,\omega_2\}
\quad \text{and} \quad
\{\omega_3,\omega_4\},
$$

the random variable $\mathbb{E}[X\mid Y]$ must take a single, constant value on each of these respective sets.

Consequently, we express the conditional expectation as:

$$
\mathbb{E}[X\mid Y](\omega)=
\begin{cases}
a, & \omega\in\{\omega_1,\omega_2\},\\
b, & \omega\in\{\omega_3,\omega_4\},
\end{cases}
$$

where $a$ and $b$ are constants to be explicitly determined.

---

### 4)
To find the constants $a$ and $b$, we use the defining integration property of conditional expectation across the generating blocks of $\sigma(Y)$.

On the set $\{\omega_1,\omega_2\}$,

$$
\int_{\{\omega_1,\omega_2\}} \mathbb{E}[X\mid Y]\,dP=\int_{\{\omega_1,\omega_2\}} X\,dP.
$$

The left-hand side simplifies to:

$$
a\cdot P(\{\omega_1,\omega_2\})=a(0.50+0.10)=0.60a.
$$

The right-hand side evaluates to:

$$
X(\omega_1)P(\{\omega_1\}) + X(\omega_2)P(\{\omega_2\}) = 0\cdot 0.50 + 1\cdot 0.10 = 0.10.
$$

Equating the two sides gives:

$$
0.60a=0.10 \implies a=\frac{0.10}{0.60}=\frac{1}{6}.
$$

On the set $\{\omega_3,\omega_4\}$,

$$
\int_{\{\omega_3,\omega_4\}} \mathbb{E}[X\mid Y]\,dP=\int_{\{\omega_3,\omega_4\}} X\,dP.
$$

The left-hand side simplifies to:

$$
b\cdot P(\{\omega_3,\omega_4\})=b(0.20+0.20)=0.40b.
$$

The right-hand side evaluates to:

$$
X(\omega_3)P(\{\omega_3\}) + X(\omega_4)P(\{\omega_4\}) = 0\cdot 0.20 + 1\cdot 0.20 = 0.20.
$$

Equating the two sides gives:

$$
0.40b=0.20 \implies b=\frac{0.20}{0.40}=\frac{1}{2}.
$$

Therefore, writing the conditional expectation as a functional mapping of $Y$:

$$
\mathbb{E}[X\mid Y]=
\frac{1}{6} \mathbb{1}_{\{Y=0\}} + \frac{1}{2} \mathbb{1}_{\{Y=1\}}=
\begin{cases}
\dfrac{1}{6}, & Y=0,\\
\dfrac{1}{2}, & Y=1.
\end{cases}
$$

---

### 5)
We verify the foundational property $\int_A \mathbb{E}[X\mid Y]\,dP=\int_A X\,dP$ for every set $A\in\sigma(Y)$:

* For $A=\varnothing$, both integrals trivially evaluate to $0$.
* For $A=\{\omega_1,\omega_2\}$:
  $$\int_A \mathbb{E}[X\mid Y]\,dP = \frac{1}{6}\cdot P(\{\omega_1,\omega_2\}) = \frac{1}{6}\cdot 0.60 = 0.10,$$
  $$\int_A X\,dP = 0\cdot 0.50 + 1\cdot 0.10 = 0.10. \quad \checkmark$$
* For $A=\{\omega_3,\omega_4\}$:
  $$\int_A \mathbb{E}[X\mid Y]\,dP = \frac{1}{2}\cdot P(\{\omega_3,\omega_4\}) = \frac{1}{2}\cdot 0.40 = 0.20,$$
  $$\int_A X\,dP = 0\cdot 0.20 + 1\cdot 0.20 = 0.20. \quad \checkmark$$
* For $A=\Omega$:
  $$\int_\Omega \mathbb{E}[X\mid Y]\,dP = \left(\frac{1}{6}\cdot 0.60\right) + \left(\frac{1}{2}\cdot 0.40\right) = 0.10 + 0.20 = 0.30,$$
  $$\int_\Omega X\,dP = (0\cdot 0.50) + (1\cdot 0.10) + (0\cdot 0.20) + (1\cdot 0.20) = 0.30. \quad \checkmark$$

The total space verification confirms the Law of Total Expectation: $\mathbb{E}[\mathbb{E}[X\mid Y]] = \mathbb{E}[X] = 0.30$.

---

### 6)
Applying the definition of discrete conditional probability $p_{X|Y}(x \mid y) = \frac{P(X=x, Y=y)}{P(Y=y)}$, we calculate the conditional distribution of fraud across both groups:

* For $Y=0$ (Low-risk customer):
  $$p_{X|Y}(1 \mid 0) = \frac{P(X=1, Y=0)}{P(Y=0)} = \frac{P(\{\omega_2\})}{0.60} = \frac{0.10}{0.60} = \mathbf{\frac{1}{6}}$$
  $$p_{X|Y}(0 \mid 0) = \frac{P(X=0, Y=0)}{P(Y=0)} = \frac{P(\{\omega_1\})}{0.60} = \frac{0.50}{0.60} = \mathbf{\frac{5}{6}}$$

* For $Y=1$ (High-risk customer):
  $$p_{X|Y}(1 \mid 1) = \frac{P(X=1, Y=1)}{P(Y=1)} = \frac{P(\{\omega_4\})}{0.40} = \frac{0.20}{0.40} = \mathbf{\frac{1}{2}}$$
  $$p_{X|Y}(0 \mid 1) = \frac{P(X=0, Y=1)}{P(Y=1)} = \frac{P(\{\omega_3\})}{0.40} = \frac{0.20}{0.40} = \mathbf{\frac{1}{2}}$$

---

### 7)
We verify that evaluating the standard summation identity matches the results obtained via the measure-theoretic path:

* Checking at $y=0$:
  $$\sum_{x \in \{0,1\}} x \cdot p_{X|Y}(x \mid 0) = \left(0 \cdot \frac{5}{6}\right) + \left(1 \cdot \frac{1}{6}\right) = \mathbf{\frac{1}{6}} = \mathbb{E}[X\mid Y=0] \quad \checkmark$$

* Checking at $y=1$:
  $$\sum_{x \in \{0,1\}} x \cdot p_{X|Y}(x \mid 1) = \left(0 \cdot \frac{1}{2}\right) + \left(1 \cdot \frac{1}{2}\right) = \mathbf{\frac{1}{2}} = \mathbb{E}[X\mid Y=1] \quad \checkmark$$

---

### 8)
The minimum mean square error achievable by our optimal estimator is equal to the expected conditional variance of the system, defined as $\text{MSE}_{\text{min}} = \mathbb{E}[\text{Var}(X \mid Y)]$.

Since $X$ conditioned on a given risk profile is a Bernoulli variable, its conditional variance is $p(1-p)$:
$$\text{Var}(X \mid Y=0) = \frac{1}{6} \left(1 - \frac{1}{6}\right) = \frac{5}{36}$$
$$\text{Var}(X \mid Y=1) = \frac{1}{2} \left(1 - \frac{1}{2}\right) = \frac{1}{4}$$

Averaging these conditional values over the probability distribution of $Y$:
$$\mathbb{E}[\text{Var}(X \mid Y)] = \text{Var}(X \mid Y=0)P(Y=0) + \text{Var}(X \mid Y=1)P(Y=1)$$
$$\mathbb{E}[\text{Var}(X \mid Y)] = \left(\frac{5}{36} \cdot 0.60\right) + \left(\frac{1}{4} \cdot 0.40\right) = \frac{1}{12} + \frac{1}{10} = \mathbf{\frac{11}{60} \approx 0.1833}$$

---

### 9)
If the bank drops the profiling metric $Y$ entirely, the best point prediction becomes the global unconditional mean $\mathbb{E}[X] = 0.30$. The resulting benchmark tracking error corresponds to the absolute total variance of $X$:
$$\text{MSE}_{\text{global}} = \text{Var}(X) = p(1-p) = 0.30 \times (1 - 0.30) = \mathbf{0.21}$$

The Law of Total Variance mathematically structures this relationship by splitting the system's baseline variance into two components:
$$\text{Var}(X) = \mathbb{E}[\text{Var}(X \mid Y)] + \text{Var}(\mathbb{E}[X \mid Y])$$

Calculating the variance explained by our conditioning estimator, $\text{Var}(\mathbb{E}[X \mid Y])$:
$$\text{Var}(\mathbb{E}[X \mid Y]) = \mathbb{E}[(\mathbb{E}[X \mid Y])^2] - (\mathbb{E}[X])^2$$
$$\text{Var}(\mathbb{E}[X \mid Y]) = \left[ \left(\frac{1}{6}\right)^2(0.60) + \left(\frac{1}{2}\right)^2(0.40) \right] - (0.30)^2 = \frac{7}{60} - 0.09 \approx \mathbf{0.0267}$$

The complete structural breakdown aligns exactly:
$$0.21 \, (\text{Total System Error}) = 0.1833 \, (\text{Unexplained Model Noise}) + 0.0267 \, (\text{Explained Informational Gain})$$

Using $Y$ directly shrinks the bank's predictive error from $0.21$ down to $0.1833$.

---

### 10)
Based on the joint mapping profiles, the system's marginal parameters are:
* Marginal for $X$: $P(X=0) = 0.70, \quad P(X=1) = 0.30$
* Marginal for $Y$: $P(Y=0) = 0.60, \quad P(Y=1) = 0.40$

**a)** The baseline background uncertainty before profile categorization is:
$$H(X) = - \left(0.70 \log_2 0.70 + 0.30 \log_2 0.30\right) \approx \mathbf{0.8813 \text{ bits}}$$

**b)** The average residual uncertainty remaining after isolating the risk group is:
$$H(X \mid Y=0) = - \left(\frac{5}{6}\log_2 \frac{5}{6} + \frac{1}{6}\log_2 \frac{1}{6}\right) \approx 0.6500 \text{ bits}$$
$$H(X \mid Y=1) = - \left(\frac{1}{2}\log_2 \frac{1}{2} + \frac{1}{2}\log_2 \frac{1}{2}\right) = 1.0000 bit$$
$$H(X \mid Y) = P(Y=0)H(X \mid Y=0) + P(Y=1)H(X \mid Y=1)$$
$$H(X \mid Y) = (0.60 \times 0.6500) + (0.40 \times 1.0000) = \mathbf{0.7900 \text{ bits}}$$

**c)** The absolute mutual information transferred via the risk feature is:
$$I(X; Y) = H(X) - H(X \mid Y) = 0.8813 - 0.7900 = \mathbf{0.0913 \text{ bits}}$$

**d)** The relative percentage of uncertainty eliminated by the classification feature is:
$$\text{Percentage Reduction} = \frac{I(X; Y)}{H(X)} = \frac{0.0913}{0.8813} \approx \mathbf{10.36\%}$$

Knowing the customer risk class removes approximately **10.36%** of the bank

In [4]:
import numpy as np

# Setup arrays matching the problem definition
probs = np.array([0.50, 0.10, 0.20, 0.20])
X_Fraud = np.array([0, 1, 0, 1])
Y_Risk = np.array([0, 0, 1, 1])

# 1) Validation of Probability Mass
print(f"1) Total Probability Mass Check: {probs.sum()}")

# 4) Calculate explicit conditional expectations
p_y0 = probs[Y_Risk == 0].sum()
p_y1 = probs[Y_Risk == 1].sum()

E_X_Y0 = (X_Fraud[Y_Risk == 0] * probs[Y_Risk == 0]).sum() / p_y0
E_X_Y1 = (X_Fraud[Y_Risk == 1] * probs[Y_Risk == 1]).sum() / p_y1
print(f"4) Expected values calculated programmatically:")
print(f"   E[X | Y=0] = {E_X_Y0:.4f} (1/6)")
print(f"   E[X | Y=1] = {E_X_Y1:.4f} (1/2)")

# 8) Minimum MSE Calculation
var_x_y0 = (1/6) * (5/6)
var_x_y1 = (1/2) * (1/2)
mse_min = (var_x_y0 * p_y0) + (var_x_y1 * p_y1)
print(f"8) Minimum Model MSE: {mse_min:.4f} (11/60)")

# 9) Global MSE & Law of Total Variance Validation
E_X = (X_Fraud * probs).sum()
mse_global = (probs * ((X_Fraud - E_X)**2)).sum()
var_of_E_X_Y = (p_y0 * (E_X_Y0 - E_X)**2) + (p_y1 * (E_X_Y1 - E_X)**2)
print(f"9) Global MSE (Total Variance): {mse_global:.4f}")
print(f"   Explained Variance Component: {var_of_E_X_Y:.4f}")

# 10) Information Theoretic Outputs
p_x0, p_x1 = probs[X_Fraud == 0].sum(), probs[X_Fraud == 1].sum()
H_X = - (p_x0 * np.log2(p_x0) + p_x1 * np.log2(p_x1))
H_X_Y0 = - ((5/6)*np.log2(5/6) + (1/6)*np.log2(1/6))
H_X_Y1 = - ((1/2)*np.log2(1/2) + (1/2)*np.log2(1/2))
H_X_cond_Y = (p_y0 * H_X_Y0) + (p_y1 * H_X_Y1)
I_XY = H_X - H_X_cond_Y

print(f"10) Shannon Entropy Metrics:")
print(f"    a) H(X)   = {H_X:.4f} bits")
print(f"    b) H(X|Y) = {H_X_cond_Y:.4f} bits")
print(f"    c) I(X;Y) = {I_XY:.4f} bits")
print(f"    d) Uncertainty reduction percentage = {(I_XY / H_X) * 100:.2f}%")

1) Total Probability Mass Check: 1.0
4) Expected values calculated programmatically:
   E[X | Y=0] = 0.1667 (1/6)
   E[X | Y=1] = 0.5000 (1/2)
8) Minimum Model MSE: 0.1833 (11/60)
9) Global MSE (Total Variance): 0.2100
   Explained Variance Component: 0.0267
10) Shannon Entropy Metrics:
    a) H(X)   = 0.8813 bits
    b) H(X|Y) = 0.7900 bits
    c) I(X;Y) = 0.0913 bits
    d) Uncertainty reduction percentage = 10.36%


### QUESTION 2

### 1)
Let $\mathbf{Z} = \begin{bmatrix} \mathbf{X} \\ \mathbf{Y} \end{bmatrix} \in \mathbb{R}^{n+m}$ follow a joint multivariate normal distribution with mean vector $\boldsymbol{\mu} = \begin{bmatrix} \boldsymbol{\mu}_X \\ \boldsymbol{\mu}_Y \end{bmatrix}$ and a symmetric, positive definite covariance matrix partitioned as:

$$
\boldsymbol{\Sigma} = \begin{bmatrix} \boldsymbol{\Sigma}_{XX} & \boldsymbol{\Sigma}_{XY} \\ \boldsymbol{\Sigma}_{YX} & \boldsymbol{\Sigma}_{YY} \end{bmatrix}
$$

The joint density function $f_{\mathbf{X}, \mathbf{Y}}(\mathbf{z})$ for $\mathbf{z} \in \mathbb{R}^{n+m}$ is given by:

$$
f_{\mathbf{X}, \mathbf{Y}}(\mathbf{z}) = \frac{1}{\sqrt{(2\pi)^{n+m} |\boldsymbol{\Sigma}|}} \exp\left( -\frac{1}{2} (\mathbf{z} - \boldsymbol{\mu})^T \boldsymbol{\Sigma}^{-1} (\mathbf{z} - \boldsymbol{\mu}) \right)
$$

---

### 2)
To find the marginal and conditional profiles, we decompose the joint quadratic form $Q = (\mathbf{z} - \boldsymbol{\mu})^T \boldsymbol{\Sigma}^{-1} (\mathbf{z} - \boldsymbol{\mu})$ using the Schur complement of $\boldsymbol{\Sigma}_{YY}$ to perform a block matrix inversion. Let $\tilde{\mathbf{x}} = \mathbf{x} - \boldsymbol{\mu}_X$ and $\tilde{\mathbf{y}} = \mathbf{y} - \boldsymbol{\mu}_Y$. The inverse of the block covariance matrix can be factored as:

$$
\boldsymbol{\Sigma}^{-1} = \begin{bmatrix} \mathbf{I} & -\boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1} \\ \mathbf{0} & \mathbf{I} \end{bmatrix}^T \begin{bmatrix} \mathbf{S}_{XX}^{-1} & \mathbf{0} \\ \mathbf{0} & \boldsymbol{\Sigma}_{YY}^{-1} \end{bmatrix} \begin{bmatrix} \mathbf{I} & -\boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1} \\ \mathbf{0} & \mathbf{I} \end{bmatrix}
$$

where $\mathbf{S}_{XX} = \boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}$ is the Schur complement of $\boldsymbol{\Sigma}_{YY}$. Substituting this factorization into the quadratic form yields:

$$
Q = (\mathbf{y} - \boldsymbol{\mu}_Y)^T \boldsymbol{\Sigma}_{YY}^{-1} (\mathbf{y} - \boldsymbol{\mu}_Y) + (\mathbf{x} - \boldsymbol{\mu}^*)^T \mathbf{S}_{XX}^{-1} (\mathbf{x} - \boldsymbol{\mu}^*)
$$

where the conditional mean vector is defined as:

$$
\boldsymbol{\mu}^* = \boldsymbol{\mu}_X + \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}(\mathbf{y} - \boldsymbol{\mu}_Y)
$$

---

### 3)
Using the definition of conditional density, we have:

$$
f_{\mathbf{X}|\mathbf{Y}}(\mathbf{x} \mid \mathbf{y}) = \frac{f_{\mathbf{X}, \mathbf{Y}}(\mathbf{x}, \mathbf{y})}{f_{\mathbf{Y}}(\mathbf{y})}
$$

The marginal density of $\mathbf{Y}$ captures the first term of our exponent decomposition:

$$
f_{\mathbf{Y}}(\mathbf{y}) = \frac{1}{\sqrt{(2\pi)^m |\boldsymbol{\Sigma}_{YY}|}} \exp\left( -\frac{1}{2} (\mathbf{y} - \boldsymbol{\mu}_Y)^T \boldsymbol{\Sigma}_{YY}^{-1} (\mathbf{y} - \boldsymbol{\mu}_Y) \right)
$$

Dividing the joint density by this marginal density cancels out the $\mathbf{Y}$-marginal quadratic term from the exponential expression:

$$
Q_{\text{conditional}} = Q_{\text{joint}} - Q_{\text{marginal}} = (\mathbf{x} - \boldsymbol{\mu}^*)^T \mathbf{S}_{XX}^{-1} (\mathbf{x} - \boldsymbol{\mu}^*)
$$

For the coefficient factors, dividing the joint constant by the marginal constant gives:

$$
\frac{\frac{1}{\sqrt{(2\pi)^{n+m} |\boldsymbol{\Sigma}|}}}{\frac{1}{\sqrt{(2\pi)^m |\boldsymbol{\Sigma}_{YY}|}}} = \frac{1}{\sqrt{(2\pi)^n \frac{|\boldsymbol{\Sigma}|}{|\boldsymbol{\Sigma}_{YY}|}}}
$$

Using the block determinant identity $|\boldsymbol{\Sigma}| = |\boldsymbol{\Sigma}_{YY}| \cdot |\mathbf{S}_{XX}|$, the fraction inside the root simplifies directly to $|\mathbf{S}_{XX}|$. Combining the simplified constant fraction and the remaining exponent profile yields the exact target conditional density function:

$$
f_{\mathbf{X}|\mathbf{Y}}(\mathbf{x} \mid \mathbf{y}) = \frac{1}{\sqrt{(2\pi)^n |\mathbf{S}_{XX}|}} \exp\left( -\frac{1}{2} (\mathbf{x} - \boldsymbol{\mu}^*)^T \mathbf{S}_{XX}^{-1} (\mathbf{x} - \boldsymbol{\mu}^*) \right)
$$

This matches the density function of a multivariate normal distribution:

$$
(\mathbf{X} \mid \mathbf{Y} = \mathbf{y}) \sim \mathcal{N}\left(\boldsymbol{\mu}_X + \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}(\mathbf{y} - \boldsymbol{\mu}_Y), \; \boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}\right)
$$

---

### 4)
By the Disintegration Theorem, the conditional expectation function $g(\mathbf{y}) = \mathbb{E}[\mathbf{X} \mid \mathbf{Y} = \mathbf{y}]$ is calculated as the first moment (mean) against the conditional probability density function:

$$
g(\mathbf{y}) = \int_{\mathbb{R}^n} \mathbf{x} f_{\mathbf{X}|\mathbf{Y}}(\mathbf{x} \mid \mathbf{y}) \, d\mathbf{x}
$$

Since $f_{\mathbf{X}|\mathbf{Y}}(\mathbf{x} \mid \mathbf{y})$ is a valid Gaussian density centered at the shifting location parameter $\boldsymbol{\mu}^*$, evaluating this moment integral yields the mean parameter directly:

$$
g(\mathbf{y}) = \boldsymbol{\mu}(\mathbf{y}) = \boldsymbol{\mu}_X + \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}(\mathbf{y} - \boldsymbol{\mu}_Y)
$$

Substituting the random vector $\mathbf{Y}$ back into our functional mapping gives the final system conditional expectation:

$$
\mathbb{E}[\mathbf{X} \mid \mathbf{Y}] = \boldsymbol{\mu}_X + \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}(\mathbf{Y} - \boldsymbol{\mu}_Y)
$$

---

### 5)
The conditional covariance matrix corresponds to the variance profile of our conditional distribution, which is equal to the constant Schur block:

$$
\text{Var}(\mathbf{X} \mid \mathbf{Y} = \mathbf{y}) = \boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}
$$

Because this matrix is completely independent of the specific realized parameter value $\mathbf{y}$, taking its expected value across the outer state space leaves the terms completely unchanged:

$$
\mathbb{E}[\text{Var}(\mathbf{X} \mid \mathbf{Y})] = \boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}
$$

Applying the Law of Total Variance to check our structural alignment:

$$
\text{Var}(\mathbf{X}) = \mathbb{E}[\text{Var}(\mathbf{X} \mid \mathbf{Y})] + \text{Var}(\mathbb{E}[\mathbf{X} \mid \mathbf{Y}])
$$

Substituting our derived profiles into the right-hand side expressions confirms the breakdown:

$$
\text{Var}(\mathbb{E}[\mathbf{X} \mid \mathbf{Y}]) = \text{Var}\left(\boldsymbol{\mu}_X + \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}(\mathbf{Y} - \boldsymbol{\mu}_Y)\right) = \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX} = \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}
$$

Summing the components together yields:

$$
\left(\boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}\right) + \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX} = \boldsymbol{\Sigma}_{XX} = \text{Var}(\mathbf{X}) \quad \checkmark
$$

The terms cancel perfectly, showing that $\boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}$ acts as the **Explained Variance**, while the Schur complement matrix represents the **Unexplained Variance (Error)**.

---

### 6)
The differential entropy of an arbitrary $d$-dimensional multivariate normal random variable $\mathbf{V} \sim \mathcal{N}(\boldsymbol{\mu}_V, \boldsymbol{\Sigma}_V)$ is given by the standard formula:

$$
h(\mathbf{V}) = \frac{1}{2}\ln\left((2\pi e)^d |\boldsymbol{\Sigma}_V|\right) = \frac{d}{2}(1 + \ln(2\pi)) + \frac{1}{2}\ln|\boldsymbol{\Sigma}_V|
$$

Applying this formula to our system components:

**a)** The primary background uncertainty of the total concatenated random matrix $\mathbf{Z} \in \mathbb{R}^{n+m}$ is:
$$h(\mathbf{Z}) = \frac{1}{2}\ln\left((2\pi e)^{n+m} |\boldsymbol{\Sigma}|\right)$$

**b)** The residual system uncertainty profile of $\mathbf{X}$ after observing the context vector $\mathbf{Y}$ is governed by the conditional covariance matrix $\mathbf{S}_{XX}$:
$$h(\mathbf{X} \mid \mathbf{Y}) = \frac{1}{2}\ln\left((2\pi e)^n |\mathbf{S}_{XX}|\right) = \frac{1}{2}\ln\left((2\pi e)^n |\boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}|\right)$$

**c)** The mutual information transmission $I(\mathbf{X}; \mathbf{Y})$ measuring the dependency between the spaces is:
$$I(\mathbf{X}; \mathbf{Y}) = h(\mathbf{X}) - h(\mathbf{X} \mid \mathbf{Y})$$
$$I(\mathbf{X}; \mathbf{Y}) = \frac{1}{2}\ln\left((2\pi e)^n |\boldsymbol{\Sigma}_{XX}|\right) - \frac{1}{2}\ln\left((2\pi e)^n |\boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}|\right)$$
$$I(\mathbf{X}; \mathbf{Y}) = \frac{1}{2}\ln\left( \frac{|\boldsymbol{\Sigma}_{XX}|}{|\boldsymbol{\Sigma}_{XX} - \boldsymbol{\Sigma}_{XY}\boldsymbol{\Sigma}_{YY}^{-1}\boldsymbol{\Sigma}_{YX}|} \right)$$

In [5]:
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# 1. Define foundational parameters for a 2D Gaussian system
mu = np.array([0.0, 0.0])
# Construct a positive definite block covariance matrix
sigma = np.array([[1.0, 0.7],
                  [0.7, 1.0]])

# Extract the individual scalar parameter blocks
sigma_xx = sigma[0, 0]
sigma_yy = sigma[1, 1]
sigma_xy = sigma[0, 1]
sigma_yx = sigma[1, 0]

# Generate synthetic data points from the joint space
n_samples = 5000
np.random.seed(42)
data = np.random.multivariate_normal(mu, sigma, n_samples)
x_samples, y_samples = data[:, 0], data[:, 1]

# 2. Compute theoretical conditional updates given a fixed observation profile
y_fixed = 1.2
mu_cond = mu[0] + (sigma_xy / sigma_yy) * (y_fixed - mu[1])
var_cond = sigma_xx - (sigma_xy**2 / sigma_yy)

# Filter samples where Y falls within a tight window near our target slice
condition_mask = np.abs(y_samples - y_fixed) < 0.1
x_conditional_samples = x_samples[condition_mask]

# 3. Compute Information-Theoretic Metrics
h_x = 0.5 * np.log(2 * np.pi * np.e * sigma_xx)
h_x_given_y = 0.5 * np.log(2 * np.pi * np.e * var_cond)
mutual_info = 0.5 * np.log(sigma_xx / var_cond)

print("="*60)
print("SYSTEM NUMERICAL VERIFICATION VERDICT")
print("="*60)
print(f"Conditional Distribution Target Center (mu*): {mu_cond:.4f}")
print(f"Conditional Distribution System Variance (S_xx): {var_cond:.4f}")
print("-"*60)
print(f"Unconditional System Entropy h(X)   : {h_x:.4f} nats")
print(f"Residual Conditional Entropy h(X|Y) : {h_x_given_y:.4f} nats")
print(f"Transferred Mutual Information I(X;Y): {mutual_info:.4f} nats")
print("="*60)

# 4. Generate Interactive Visualization Layout
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("Joint Density Space P(X, Y)", "Marginal Profile P(Y)",
                    "Marginal Profile P(X)", f"Conditional Profile P(X | Y={y_fixed})")
)

# Top Left: Joint Scattering
fig.add_trace(go.Scatter(x=x_samples[::5], y=y_samples[::5], mode='markers',
                         marker=dict(color='blue', opacity=0.3, size=3), name="Joint Space"), row=1, col=1)
fig.add_shape(type="line", x0=-3, y0=y_fixed, x1=3, y1=y_fixed,
              line=dict(color="Red", width=2, dash="dash"), row=1, col=1)

# Top Right: Marginal Y Histogram
fig.add_trace(go.Histogram(y=y_samples, nbinsy=40, marker_color='green', name="Marginal Y"), row=1, col=2)

# Bottom Left: Marginal X Histogram
fig.add_trace(go.Histogram(x=x_samples, nbinsx=40, marker_color='orange', name="Marginal X"), row=2, col=1)

# Bottom Right: Shifting Conditional Distribution Profile Histogram
fig.add_trace(go.Histogram(x=x_conditional_samples, nbinsx=30, marker_color='red',
                           histnorm='probability density', name="Empirical X|Y"), row=2, col=2)

# Overlay theoretical Gaussian curve onto the empirical slice histogram
x_axis = np.linspace(-3, 3, 200)
y_gauss = (1.0 / np.sqrt(2 * np.pi * var_cond)) * np.exp(-0.5 * ((x_axis - mu_cond)**2) / var_cond)
fig.add_trace(go.Scatter(x=x_axis, y=y_gauss, mode='lines', line=dict(color='black', width=3), name="Theoretical PDF"), row=2, col=2)

fig.update_layout(title_text="Multivariate Normal Conditioning Analysis", height=700, width=900, showlegend=False)
fig.show()

SYSTEM NUMERICAL VERIFICATION VERDICT
Conditional Distribution Target Center (mu*): 0.8400
Conditional Distribution System Variance (S_xx): 0.5100
------------------------------------------------------------
Unconditional System Entropy h(X)   : 1.4189 nats
Residual Conditional Entropy h(X|Y) : 1.0823 nats
Transferred Mutual Information I(X;Y): 0.3367 nats


### QUESTION 3

### 1)
Since $X \sim \mathcal{N}(0, 1)$ and $Q \sim \mathcal{N}(0, 4)$ are independent normal random variables, any linear combination of $X$ and $Y = X + Q$ is also a linear combination of the independent normals $X$ and $Q$. Therefore, the vector $(X, Y)^T$ follows a joint multivariate normal distribution:

$$
\begin{bmatrix} X \\ Y \end{bmatrix} \sim \mathcal{N}(\boldsymbol{\mu}, \boldsymbol{\Sigma})
$$

To find the marginal distribution of $Y$, we look at its expectation and variance:
* **Mean:** $\mathbb{E}[Y] = \mathbb{E}[X + Q] = \mathbb{E}[X] + \mathbb{E}[Q] = 0 + 0 = 0$
* **Variance:** Since $X$ and $Q$ are independent, $\text{Var}(Y) = \text{Var}(X + Q) = \text{Var}(X) + \text{Var}(Q) = 1 + 4 = 5$

Thus, $Y \sim \mathcal{N}(0, 5)$.

To find the covariance $\text{Cov}(X, Y) = \Sigma_{XY}$:

$$
\Sigma_{XY} = \mathbb{E}[(X - \mathbb{E}[X])(Y - \mathbb{E}[Y])] = \mathbb{E}[X(X + Q)] = \mathbb{E}[X^2] + \mathbb{E}[XQ]
$$

Using independence ($\mathbb{E}[XQ] = \mathbb{E}[X]\mathbb{E}[Q] = 0$) and the definition of variance for a zero-mean variable ($\mathbb{E}[X^2] = \text{Var}(X) = 1$):

$$
\Sigma_{XY} = 1 + 0 = 1
$$

The complete joint parameters are:

$$
\boldsymbol{\mu} = \begin{bmatrix} 0 \\ 0 \end{bmatrix}, \qquad \boldsymbol{\Sigma} = \begin{bmatrix} 1 & 1 \\ 1 & 5 \end{bmatrix}
$$

---

### 2)
A random variable $X$ is $\sigma(Y)$-measurable if and only if its value can be uniquely determined from a realization of $Y$. Mathematically, $\sigma(Y)$ contains information regarding the sum $X + Q$, which creates a pool of infinite combinations of $X$ and $Q$ that can yield the exact same value for $Y$.

In practical terms, if the power meter shows $Y = 5$, the engineer cannot isolate the true temperature because multiple physical scenarios map to this exact observation:
* **Overheating scenario:** The spindle could be dangerously hot ($X = 4$) with a standard mechanical cutting load ($Q = 1$).
* **Hard material scenario:** The spindle could be running at its optimal baseline temperature ($X = 0$) but cutting an exceptionally hard batch of material ($Q = 5$).

Because the mapping from $Y \to X$ is one-to-many, $X$ is not $\sigma(Y)$-measurable, creating fundamental informational ambiguity.

---

### 3)
Using the explicit projection property for joint multivariate normal distributions, the conditional expectation is:

$$
\mathbb{E}[X \mid Y = y] = \mu_X + \frac{\text{Cov}(X,Y)}{\text{Var}(Y)}(y - \mu_Y)
$$

Substituting our known system parameters ($\mu_X = 0$, $\mu_Y = 0$, $\text{Cov}(X,Y) = 1$, $\text{Var}(Y) = 5$):

$$
\mathbb{E}[X \mid Y = y] = 0 + \frac{1}{5}(y - 0) = \frac{1}{5}y
$$

Expressing this as a functional random variable mapping:

$$
\mathbb{E}[X \mid Y] = \frac{1}{5}Y
$$

This expression reflects a "signal-to-total-variance" ratio. By rewriting the denominator using our structural blocks:

$$
\mathbb{E}[X \mid Y] = \left( \frac{\text{Var}(X)}{\text{Var}(X) + \text{Var}(Q)} \right) Y = \left( \frac{1}{1 + 4} \right) Y = \frac{1}{5}Y
$$

The weight assigned to our observation is directly proportional to how much of the total monitored power variance is driven by true thermal signal versus background mechanical noise.

---

### 4)
Let our optimal estimate be $\widehat{X} = \frac{1}{5}Y$. The estimation residual is defined as:

$$
Z = X - \widehat{X} = X - \frac{1}{5}Y
$$

We check the unbiased property by taking the expectation:

$$
\mathbb{E}[Z] = \mathbb{E}\left[X - \frac{1}{5}Y\right] = \mathbb{E}[X] - \frac{1}{5}\mathbb{E}[Y] = 0 - \frac{1}{5}(0) = 0
$$

To prove that the error is uncorrelated with our monitoring variable $Y$ (Orthogonality Principle), we compute their covariance:

$$
\text{Cov}(Z, Y) = \text{Cov}\left(X - \frac{1}{5}Y, \; Y\right) = \text{Cov}(X, Y) - \frac{1}{5}\text{Cov}(Y, Y)
$$

Substituting the variance of $Y$ ($\text{Cov}(Y,Y) = \text{Var}(Y) = 5$) and our joint covariance ($\text{Cov}(X,Y) = 1$):

$$
\text{Cov}(Z, Y) = 1 - \frac{1}{5}(5) = 1 - 1 = 0
$$

Since their covariance is zero, the residual error $Z$ is perfectly uncorrelated with the observation $Y$. Any informational patterns present in $Y$ have been completely extracted by the estimator.

---

### 5)
The differential entropy of a continuous univariate normal variable $V \sim \mathcal{N}(\mu_V, \sigma_V^2)$ is $h(V) = \frac{1}{2}\ln(2\pi e \sigma_V^2)$.

* **Unconditional Entropy of $X$:**
  Since $\sigma_X^2 = 1$:
  $$h(X) = \frac{1}{2}\ln(2\pi e (1)) = \frac{1}{2}\ln(2\pi e) \approx 1.4189 \text{ nats}$$

* **Conditional Entropy $h(X \mid Y)$:**
  The variance of the conditional distribution is the remaining uncertainty left in the Schur complement:
  $$\sigma_{X|Y}^2 = \sigma_X^2 - \frac{\text{Cov}(X,Y)^2}{\text{Var}(Y)} = 1 - \frac{1^2}{5} = 1 - 0.20 = 0.80$$
  
  Evaluating the entropy formula with this reduced conditional variance yields:
  $$h(X \mid Y) = \frac{1}{2}\ln(2\pi e \sigma_{X|Y}^2) = \frac{1}{2}\ln(2\pi e (0.80)) \approx 1.3073 \text{ nats}$$

---

### 6)
The mutual information content shared between the state vector and the measurement space is:

$$
I(X; Y) = h(X) - h(X \mid Y) = \frac{1}{2}\ln(2\pi e) - \frac{1}{2}\ln(2\pi e \cdot 0.80) = \frac{1}{2}\ln\left(\frac{1}{0.80}\right) = \frac{1}{2}\ln(1.25) \approx 0.1116 \text{ nats}
$$

* **Percentage of Uncertainty Removed:**
  The percentage of internal entropy eliminated by using the power meter is:
  $$\text{Percentage Reduction} = \frac{I(X; Y)}{h(X)} = \frac{\frac{1}{2}\ln(1.25)}{\frac{1}{2}\ln(2\pi e)} = \frac{0.2231}{2.8379} \approx \mathbf{7.86\%}$$

* **Impact of High Mechanical Noise ($\text{Var}(Q) = 100$):**
  If the mechanical noise increases to $\text{Var}(Q) = 100$, the new variance of the power meter becomes $\text{Var}(Y) = 1 + 100 = 101$.
  
  The new conditional variance becomes:
  $$\sigma_{X|Y, \; \text{new}}^2 = 1 - \frac{1^2}{101} = \frac{100}{101} \approx 0.9901$$
  
  The new mutual information transfer drops significantly to:
  $$I(X; Y)_{\text{new}} = \frac{1}{2}\ln\left(\frac{1}{0.9901}\right) = \frac{1}{2}\ln(1.01) \approx \mathbf{0.0050 \text{ nats}}$$
  
  The new percentage reduction drops to:
  $$\text{Percentage Reduction}_{\text{new}} = \frac{\frac{1}{2}\ln(1.01)}{\frac{1}{2}\ln(2\pi e)} \approx \mathbf{0.35\%}$$
  
  When torque variance scales up, the measurement becomes flooded with mechanical noise, rendering the power meter nearly useless for tracking internal temperature anomalies.

---

### 7)
The manager's claim that $\widehat{X} = Y$ is a good approach overlooks the scaling of the noise term. If we evaluate the mean squared error of the manager's estimator, we find:

$$
\mathbb{E}[(Y - X)^2] = \mathbb{E}[Q^2] = \text{Var}(Q) = 4
$$

In contrast, our conditional expectation estimator yields an error equal to the conditional variance:

$$
\mathbb{E}[(\mathbb{E}[X\mid Y] - X)^2] = \sigma_{X|Y}^2 = 0.80
$$

The raw measurement $Y$ is an over-responsive estimator because it attributes every fluctuation in power directly to temperature changes.

The optimal minimum mean squared error (MMSE) estimator, $\mathbb{E}[X \mid Y] = \frac{1}{5}Y$, explicitly "shrinks" the measurement toward the global mean (zero). This shrinkage is a deliberate design feature dictated by the signal-to-noise ratio. Because $80\%$ of the variance in the power reading is driven by unobserved mechanical torque ($Q$), any large spike in $Y$ is statistically more likely to be caused by a heavy mechanical load than by a thermal spike. The estimator accounts for this by scaling back the measurement, protecting the system from generating false alarms due to normal torque variations.

In [ ]:
import numpy as np

# Define problem parameters
var_x = 1.0
var_q = 4.0
var_y = var_x + var_q  # 1 + 4 = 5
cov_xy = var_x         # 1.0

print("="*60)
print("CNC SPINDLE DIAGNOSTIC VERIFICATION")
print("="*60)

# 3) Verification of Signal-to-Noise Coefficient
coefficient = cov_xy / var_y
print(f"3) Optimal Estimator Equation: E[X | Y] = {coefficient:.2f} * Y")

# 4) Simulation-Based Proof of Orthogonality Principle
np.random.seed(42)
X = np.random.normal(0, np.sqrt(var_x), 100000)
Q = np.random.normal(0, np.sqrt(var_q), 100000)
Y = X + Q

X_hat = coefficient * Y
Z = X - X_hat

print(f"4) Simulation Results (100k samples):")
print(f"   Expected Error E[Z]      = {np.mean(Z):.4f} (Theoretical: 0)")
print(f"   Covariance Cov(Z, Y)     = {np.cov(Z, Y)[0, 1]:.4f} (Theoretical: 0)")

# 5 & 6) Information Theory Calculations
h_x = 0.5 * np.log(2 * np.pi * np.e * var_x)
var_cond = var_x - (cov_xy**2 / var_y)
h_x_cond_y = 0.5 * np.log(2 * np.pi * np.e * var_cond)
I_xy = h_x - h_x_cond_y
pct_reduction = (I_xy / h_x) * 100

print(f"5) Differential Entropies:")
print(f"   h(X)   = {h_x:.4f} nats")
print(f"   h(X|Y) = {h_x_cond_y:.4f} nats (Conditional Variance: {var_cond:.2f})")
print(f"6) Information Gain:")
print(f"   Mutual Information I(X;Y) = {I_xy:.4f} nats")
print(f"   Uncertainty Reduction     = {pct_reduction:.2f}%")

# High Noise Variance Case (Var(Q) = 100)
var_q_high = 100.0
var_y_high = var_x + var_q_high
var_cond_high = var_x - (cov_xy**2 / var_y_high)
h_x_cond_y_high = 0.5 * np.log(2 * np.pi * np.e * var_cond_high)
I_xy_high = h_x - h_x_cond_y_high
pct_reduction_high = (I_xy_high / h_x) * 100

print(f"   [With Var(Q)=100] New I(X;Y) = {I_xy_high:.4f} nats")
print(f"   [With Var(Q)=100] New Reduction = {pct_reduction_high:.2f}%")
print("="*60)